python-oauth-web-app/app.py

```diff
@@ -0,0 +1,300 @@
+import time
+import sqlite3
+import functools
+from datetime import datetime, timezone
+from urllib.parse import urlparse, parse_qs, urlencode
+from flask import (
+    Flask,
+    flash,
+    url_for,
+    redirect,
+    render_template,
+    jsonify,
+    request,
+    g,
+    session,
+    abort,
+)
+from authlib.jose import JsonWebKey
+from authlib.common.security import generate_token
+from authlib.jose import jwt
+from authlib.oauth2.rfc7636 import create_s256_code_challenge
+
+from atproto_identity import *
+from atproto_oauth import *
+from atproto_security import is_safe_url, hardened_http
+
+app = Flask(__name__)
+app.config.from_prefixed_env()
+
+secret_jwk = JsonWebKey.import_key(app.config["SECRET_JWK"])
+pub_jwk = json.loads(secret_jwk.as_json(is_private=False))
+assert "d" not in pub_jwk
+
+
+def get_db():
+    db = getattr(g, "_database", None)
+    if db is None:
+        db_path = app.config.get("DATABASE_URL", "demo.sqlite")
+        db = g._database = sqlite3.connect(db_path)
+        db.row_factory = sqlite3.Row
+    return db
+
+
+def query_db(query, args=(), one=False):
+    conn = get_db()
+    cur = conn.cursor()
+    cur.execute(query, args)
+    rv = cur.fetchall()
+    conn.commit()
+    cur.close()
+    return (rv[0] if rv else None) if one else rv
+
+
+def init_db():
+    print("initializing database...")
+    with app.app_context():
+        db = get_db()
+        with app.open_resource("schema.sql", mode="r") as f:
+            db.cursor().executescript(f.read())
+        db.commit()
+
+
+init_db()
+
+
+@app.teardown_appcontext
+def close_connection(exception):
+    db = getattr(g, "_database", None)
+    if db is not None:
+        db.close()
+
+
+@app.before_request
+def load_logged_in_user():
+    user_did = session.get("user_did")
```

Comment:
Note that you might be able to impersonate any user just by setting their `did` in your cookie !

Reply:
Or is the cookie secure? (then maybe add a comment ?)